src/clj_foundation/errors.clj

```diff
@@ -1,23 +1,37 @@
 (ns clj-foundation.errors
   (:require [clojure.tools.logging :as log]
+            [clj-foundation.patterns :refer :all]
             [schema.core :as s :refer [=> =>*]])
 
   (:import [java.util Date])
 
   (:gen-class))
 
 
-(defn replace-nil
-  [maybe-nil replacement]
-  (if (nil? maybe-nil)
-    replacement
-    maybe-nil))
+(s/defn exception-seq :- [Throwable]
+  "Returns a lazy sequence of all nested exceptions starting with the top-level exception and working to
+   the initial exception by recursively unrolling the (.getCause exception) for each nested exception.."
+  [exception :- Throwable]
+  (if exception
+    (lazy-seq (cons exception (exception-seq (.getCause exception))))))
 
 
-(defn not-nil [value name]
-  (if (nil? value)
-    (throw (java.lang.IllegalArgumentException. (str name " cannot be nil")))
-    value))
+(s/defn expect-within :- s/Any
+  "Expect the condition specified by predicate to become true within timeout-millis. If this
+  does not happen, throws IllegalStateException including the error-message.  On success, returns
+  the truthy value that predicate returned."
+  [timeout-millis :- s/Num, predicate :- (=> s/Any []), error-message :- s/Str]
+
+  (let [before (.getTime (Date.))]
+    (loop [completed (predicate)]
+      (let [later (.getTime (Date.))]
+        (if (< timeout-millis (- later before))
+          (throw (IllegalStateException. error-message))
+          (if-not completed
+            (do
+              (Thread/sleep 500)
+              (recur (predicate)))
```

Comment:
From the clojure reference for recur:

Evaluates the exprs in order, then, in parallel, rebinds the bindings of the recursion point to the values of the exprs. If the recursion point was a fn method, then it rebinds the params. If the recursion point was a loop, then it rebinds the loop bindings. Execution then jumps back to the recursion point.

So, rebinding of the (loop [args]) happens in parallel, but the body runs sequentially (and via tail-call optimization) in a loop.